rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,770 @@
+        return rv.arg1
+
+    @ContextMixin.use_contextual_arguments
+    def sdram_alloc_as_io(self, size, tag=0, x=Required, y=Required,
+                          app_id=Required):
+        """Like :py:meth:`.sdram_alloc` but a file-like object which allows
+        reading and writing to the region is returned.
+
+        Returns
+        -------
+        :py:class:`~MemoryIO`
+            File-like object which allows accessing the newly allocated region
+            of memory.
+
+        Raises
+        ------
+        SpiNNakerMemoryError
+            If the memory cannot be allocated, or the tag is already taken or
+            invalid.
+        """
+        # Perform the malloc
+        start_address = self.sdram_alloc(size, tag, x, y, app_id)
+
+        return MemoryIO(self, x, y, start_address, start_address + size)
+
+    def _get_next_nn_id(self):
+        """Get the next nearest neighbour ID."""
+        self._nn_id = self._nn_id + 1 if self._nn_id < 126 else 1
+        return self._nn_id * 2
+
+    def _send_ffs(self, pid, region, n_blocks, fr):
+        """Send a flood-fill start packet."""
+        sfr = fr | (1 << 31)
+        self._send_scp(
+            0, 0, 0, SCPCommands.nearest_neighbour_packet,
+            (NNCommands.flood_fill_start << 24) | (pid << 16) |
+            (n_blocks << 8),
+            region, sfr
+        )
+
+    def _send_ffd(self, pid, aplx_data, address):
+        """Send flood-fill data packets."""
+        block = 0
+        while len(aplx_data) > 0:
+            # Get the next block of data, send and progress the block
+            # counter and the address
+            data, aplx_data = (aplx_data[:consts.SCP_DATA_LENGTH],
+                               aplx_data[consts.SCP_DATA_LENGTH:])
+            size = len(data) // 4 - 1
+
+            arg1 = (NNConstants.forward << 24 | NNConstants.retry << 16 | pid)
+            arg2 = (block << 16) | (size << 8)
+            self._send_scp(0, 0, 0, SCPCommands.flood_fill_data,
+                           arg1, arg2, address, data)
+
+            # Increment the address and the block counter
+            block += 1
+            address += len(data)
+
+    def _send_ffe(self, pid, app_id, app_flags, cores, fr):
+        """Send a flood-fill end packet."""
+        arg1 = (NNCommands.flood_fill_end << 24) | pid
+        arg2 = (app_id << 24) | (app_flags << 18) | (cores & 0x3fff)
+        self._send_scp(0, 0, 0, SCPCommands.nearest_neighbour_packet,
+                       arg1, arg2, fr)
+
+    @ContextMixin.require_named_contextual_arguments("app_id", "app_flags")
+    def load_aplx(self, *args, **kwargs):
+        """Load an APLX to a set of application cores.
+
+        If a :py:class:`str` APLX filename is the first argument then the
+        second is assumed to be a dictionary mapping {(x, y): set([cores]),
+        ...}.  Otherwise the return value of
+        :py:func:`~rig.place_and_route.util.build_application_map` may be used
+        directly.
```

Comment:
.. warning::
        The loading process is extremely likely, but not guaranteed, to succeed.
        This is because the flood-fill packets used during loading are not
        guaranteed to arrive. The effect of this is one of the following:
        
        * Some regions may be included/excluded incorrectly.
        * Some chips will not receive the complete application binary and will
          silently not execute the binary.
        
        As a result, the user is responsible for checking that each core was
        successfully loaded with the correct binary. At present, the two
        recommended approaches to this are:
        
        * The user should check that the correct number of application binaries
          reach their initial barrier (SYNC0), when this facility is used. This
          is not fool-proof but will flag up all but situations where exactly
          the right number, but the wrong selection of cores were loaded. (At
          the time of writing, this situation is not possible but will become
          a concern in future versions of SC&MP.
        * The user can check the process list of each chip to ensure the
          application was loaded into the correct set of cores.

This warning obviously will need to be re-written once checking becomes part of this method.